Mason: Resource Provisionning

sebastienvas opened the pull request:
Mason is used to setup resources from existing (leased) resources. As an example, we can create a cluster from a GCP project. Mason takes care of acquiring the resources, and updating the new resource. Each vendor can implement their own "Masonable". I added a simple implementation for GKE cluster and VM on GCP.

More details at https://goo.gl/vHNfww

- Updates boskos Makefile to build mason
- Updates resources.json to yaml for readibility
- Adds AcquireByState method to Ranch, Boskos and client, to be able to Acquire multiple (leased) resources at once
- Fix linter and gometalinter issues in boskos directory
- Updates Acquire to return a resources instead of a name
- Added support for user data to Resources
- Separated ranch functionality from storage for re-usability and ability to swap storage layer
- Implemented custom resource definition storage for resources (and other)
- Mason implementation + GCP implementation
- Create a Mason client to ease acquisition and release of leased resources.

sebastienvas commented:
/assign @krzyzacy

BenTheElder commented:
have you run `hack/update-deps.sh`?  PR size suggests vendor is not pruned (?)

BenTheElder commented:
/ok-to-test
/area boskos

sebastienvas commented:
@BenTheElder that s how I updated deps. Very cool tool.

BenTheElder commented:
> @BenTheElder that s how I updated deps. Very cool tool.

@fejta :-) ^

So... that means you're actually vendoring ~300k lines? 😱 :this_is_fine:
/shrug

sebastienvas commented:
Yes those are mainly k8s :), client-go

fejta commented:
/assign @kargakis 

Michalis didn't you recently vendor the client??

BenTheElder commented:
/unshrug

We at least vendored the API machinery types, not sure about client-go though.

fejta commented:
Will you please change your first two commits to include the exact commands used to create them? Also it seems like you can combine them into a single vendor commit?

kargakis commented:
> Michalis didn't you recently vendor the client??

I vendored api and apimachinery, client-go was my next step. Unfortunately this change doesn't seem to bring in informers but that's fine for now.

sebastienvas commented:
@fejta I had them separated such that people can see that I added a new constraint and that I had to update the update-bazel script. I did not want this to be lost in the vendor update (I can still combine them, please let me know). The constraint file is updated manually. Per dep doc:
```
The Gopkg.toml file is initially generated by dep init, and is primarily hand-edited.
```
I updated the vendor update with how it was generated.

sebastienvas commented:
We get user data when we acquire the resource. I updated the docs with this. Also updated AcquireByState to take in a list of resource names. PTAL

krzyzacy commented:
(falling behind on reviews, will get onto this asap)

BenTheElder commented on boskos/boskos.go:
```diff
@@ -17,53 +17,79 @@ limitations under the License.
+
+	boskos := http.Server{
+		Handler: NewBoskosHandler(r),
+		Addr:    ":8080",
```
@krzyzacy we should probably make this configurable in the future

BenTheElder replied:
Not a nit for this PR, just a note reading back through boskos, this PR is plenty large enough already :-)

BenTheElder commented on boskos/crds/resources_config_crd.go:
```diff
@@ -0,0 +1,165 @@
+	in.Spec.Needs = r.Needs
+}
+
+// FromItem implement the Object interface
```
nit: seems like a lot of these doc comments should be s/implement/implements/

BenTheElder commented:
I took a pass at some of it, will try to dig in deeper tomorrow as well, this is a big PR 😬

BenTheElder commented on boskos/client/client.go:
```diff
@@ -17,17 +17,26 @@ limitations under the License.
-// Public Boskos client object
+var (
+	// BoskosURL defines URL where boskos HTTP server is running
+	BoskosURL = flag.String("boskos-url", "http://boskos", "Boskos URL")
```
packages really shouldn't export flags :(

sebastienvas replied:
Just out of curiosity, what's the reason ? Is there a guideline you can point me to ?

BenTheElder replied:
This registers a global flag in a flat namespace that the importing binary isn't aware of and has no control over. Usage really ought to be defined by commands, not libraries / sub-packages.

kargakis commented:
I am wondering why the needs-rebase plugin hasn't acted on this PR -- @cjwagner can you take a look?

cjwagner commented:
> I am wondering why the needs-rebase plugin hasn't acted on this PR -- @cjwagner can you take a look?

There aren't any logs indicating an error. I'm guessing that either a webhook was dropped or Github did not have the mergeability for the PR calculated (probably the later). We could update the plugin to re-request mergeability for PRs that it is not calculated for, but this could use **a lot** of API tokens.

krzyzacy commented:
I'll really do the review this afternoon... scheduled a REAL block on my calendar :-)

krzyzacy commented on boskos/mason/test-configs.yaml:
```diff
@@ -0,0 +1,23 @@
+          - http-server
+          - https-server
+          scopes:
+          - https://www.googleapis.com/auth/cloud-platform
```
nit: new line

sebastienvas replied:
done

krzyzacy commented on boskos/mason/test-resources.yaml:
```diff
@@ -0,0 +1,15 @@
+  - test-boskos-03
+  - test-boskos-04
+  - test-boskos-05
+  - test-boskos-06
```
nit: new line

sebastienvas replied:
done

krzyzacy commented on boskos/Makefile:
```diff
@@ -74,7 +75,7 @@ service:
 
 update-config: get-cluster-credentials
-	kubectl create configmap resources --from-file=config=resources.json --dry-run -o yaml | kubectl --namespace="$(NAMESPACE)" replace configmap resources -f -
+	kubectl create configmap resources --from-file=config=resources.yaml --dry-run -o yaml | kubectl --namespace="$(NAMESPACE)" replace configmap resources -f -
```
(IMO this will probably break our current deployment)

sebastienvas replied:
What do you propose we do ?

krzyzacy replied:
keep the make target with the old config.json, and make a separate PR that upgrades everything? so it will be easier to revert.

krzyzacy replied:
I think let's split the make target change in a follow up PR

krzyzacy commented on boskos/mason/mason.go:
```diff
@@ -0,0 +1,490 @@
+}
+
+// NewMason creates and initialized a new Mason object
+func NewMason(rtypes []string, channelSize, cleanerCount int, client boskosClient, sleepTime time.Duration) *Mason {
```
doc each input fields?

sebastienvas replied:
done

krzyzacy commented on hack/update-bazel.sh:
```diff
@@ -27,6 +27,10 @@ TMP_GOPATH=$(mktemp -d)
 rm -f ${TESTINFRA_ROOT}/vendor/k8s.io/apimachinery/pkg/util/sets/{BUILD,BUILD.bazel}
 
+# manually remove BUILD as testdata cannot be found
+rm -f ${TESTINFRA_ROOT}/vendor/k8s.io/client-go/util/cert/BUILD
```
cc @BenTheElder @ixdy is this fine?

BenTheElder replied:
in the future the publishing bot will start doing this last I saw

ixdy replied:
hm, is gazelle not removing the defunct `go_test`?

sebastienvas replied:
it might be now. Let me retry.

sebastienvas replied:
no this is still needed

sebastienvas replied:
```shell
$ bazel build //...            
INFO: Analysed 2077 targets (524 packages loaded).
INFO: Found 2077 targets...
ERROR: missing input file '//vendor/k8s.io/client-go/util/cert:testdata/dontUseThisKey.pem'
ERROR: /usr/local/google/home/sebastienvas/go/src/k8s.io/test-infra/boskos/BUILD.bazel:16:1: //boskos:go_default_test: missing input file '//vendor/k8s.io/client-go/util/cert:testdata/dontUseThisKey.pem'
ERROR: /usr/local/google/home/sebastienvas/go/src/k8s.io/test-infra/boskos/BUILD.bazel:16:1 1 input file(s) do not exist
INFO: Elapsed time: 6.559s, Critical Path: 0.36s
FAILED: Build did NOT complete successfully
```

krzyzacy commented:
@sebastienvas also please add a mason README to help me better understand it's functionality :-) might also link your design doc there as well?

BenTheElder commented on boskos/common/common_test.go:
```diff
@@ -0,0 +1,54 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
not if this was written in 2017

sebastienvas replied:
it was written in 2017:)

sebastienvas commented on boskos/ranch/ranch.go:
```diff
@@ -106,24 +96,84 @@ func NewRanch(config string, storage string) (*Ranch, error) {
+		res := allResources[idx]
+		if state == res.State {
+			if res.Owner != "" {
+				return nil, &OwnerNotMatch{owner, res.Owner}
```
I don't think so. In Mason use case, client should acquire all the resources in a given state. We can change this behavior, and deal with it in mason (I think we check in mason anyways).

krzyzacy replied:
from your function comment: `AcquireByState checks out resources of a given type without an owner`, so it should not return error, but continue with rest of no-owner resources, IMO

sebastienvas commented on boskos/ranch/ranch.go:
```diff
@@ -106,24 +96,84 @@ func NewRanch(config string, storage string) (*Ranch, error) {
+// and move the checked out resource to the end of the resource list.
+// In: state - current state of the requested resource
+//     dest - destination state of the requested resource
+//     owner - requester of the resource
```
done

sebastienvas commented on boskos/ranch/ranch.go:
```diff
@@ -106,24 +96,84 @@ func NewRanch(config string, storage string) (*Ranch, error) {
+					logrus.WithError(err).Errorf("could not update resource %s", res.Name)
+					return nil, err
+				}
+				resources = append(resources, res)
```
I see your point I deleted the key in the map and changed the error to list the missing resources.

sebastienvas replied:
I see your point. I deleted the key from the map instead and updated the error message with the resources missed.

sebastienvas commented on boskos/storage/storage.go:
```diff
@@ -0,0 +1,94 @@
+func (im *inMemoryStore) Add(i common.Item) error {
+	im.lock.Lock()
+	defer im.lock.Unlock()
+	im.items[i.GetName()] = i
```
done

sebastienvas commented on boskos/common/common.go:
```diff
@@ -19,37 +19,173 @@ package common
+// Update updates existing UserData with new UserData.
+// If a key as an empty string, the key will be deleted
+func (ud UserData) Update(new UserData) {
+	if new != nil {
```
I think this would be bad, and user might be deleting user data they don't know about. In order delete userdata you need to send an update of a key with an empty string. I think it is better if it s explicit.

sebastienvas commented on boskos/mason/client.go:
```diff
@@ -0,0 +1,127 @@
+				logrus.WithError(err).Warning("failed to release resource %s", name)
+				allErrors = multierror.Append(allErrors, err)
+			}
+			return
```
This is assumed as allErrors name is set in function def.

sebastienvas commented:
@krzyzacy please take a look. I ll fix the vendor and bazel tests once review is complete.

krzyzacy commented:
I'll take another pass tomorrow

krzyzacy commented on boskos/mason/mason.go:
```diff
@@ -151,6 +159,12 @@ func ValidateConfig(configs []common.ResourcesConfig, resources []common.Resourc
 // NewMason creates and initialized a new Mason object
+// In: rtypes       - A list of resource types to act on
+//     channelSize  - Size for all the channel
+// 	   cleanerCount - Number of cleaning threads
```
nit: alignment

krzyzacy commented on boskos/ranch/ranch.go:
```diff
@@ -127,6 +128,7 @@ func (r *Ranch) Acquire(rType, state, dest, owner string) (*common.Resource, err
 // In: state - current state of the requested resource
 //     dest - destination state of the requested resource
 //     owner - requester of the resource
+//     names - names of resources to acquire
```
nit: names of resource types to acquire

sebastienvas replied:
no its resource name, not type.

sebastienvas replied:
removed the "s"

krzyzacy replied:
ohhh humm then I understand this wrong - gonna walk this through again

krzyzacy commented on vendor/k8s.io/apimachinery/pkg/util/sets/BUILD.bazel:
```diff
@@ -1,29 +0,0 @@
-    srcs = [":package-srcs"],
-    tags = ["automanaged"],
-    visibility = ["//visibility:public"],
-)
```
(what's going on here?)

sebastienvas replied:
I ran hack/update-bazel.sh, I ll fix it when updating the vendors

krzyzacy commented on boskos/ranch/ranch.go:
```diff
@@ -106,24 +97,90 @@ func NewRanch(config string, storage string) (*Ranch, error) {
+// In: state - current state of the requested resource
+//     dest - destination state of the requested resource
+//     owner - requester of the resource
+//     names - names of resource to acquire
```
so `names` is actually just a count for how many resources in this state you want to fetch?

sebastienvas replied:
names is a list of resource names. So from Mason, we get the list from the leasedResources user data, and we should fetch the same with AcquireByState. But you could use this to get a subset.

krzyzacy replied:
ack :-)

sebastienvas commented on boskos/boskos.go:
```diff
@@ -131,12 +157,73 @@ func handleAcquire(r *ranch.Ranch) http.HandlerFunc {
+		if names != "" {
+			rNames = strings.Split(names, ",")
+		}
+		if state == "" || dest == "" || owner == "" {
```
there is one already in server_client_test.go

sebastienvas replied:
done

sebastienvas commented on boskos/boskos.go:
```diff
@@ -131,12 +157,73 @@ func handleAcquire(r *ranch.Ranch) http.HandlerFunc {
+			return
+		}
+
+		logrus.Infof("Request for a %v from %v, dest %v", state, owner, dest)
```
done

sebastienvas commented on boskos/boskos.go:
```diff
@@ -131,12 +157,73 @@ func handleAcquire(r *ranch.Ranch) http.HandlerFunc {
+	}
+}
+
+//  handleAcquireByState: Handler for /acquire
```
done

sebastienvas commented on boskos/boskos.go:
```diff
@@ -131,12 +157,73 @@ func handleAcquire(r *ranch.Ranch) http.HandlerFunc {
-			resource.Owner = "" // release the resource, though this is not expected to happen.
+			// release the resource, though this is not expected to happen.
+			err = r.Release(resource.Name, state, owner)
+			logrus.WithError(err).Warning("unable to release resource %s", resource.Name)
```
done

sebastienvas commented on boskos/boskos.go:
```diff
@@ -131,12 +157,73 @@ func handleAcquire(r *ranch.Ranch) http.HandlerFunc {
+			http.Error(res, err.Error(), ErrorToStatus(err))
+			for _, resource := range resources {
+				err := r.Release(resource.Name, state, owner)
+				logrus.WithError(err).Warning("unable to release resource %s", resource.Name)
```
done

krzyzacy commented:
:+1: looks fine, let's do the merge and deployment after k8s 1.10 release (that's due next Wed)

sebastienvas commented:
Cool Thanks. Let me know when you are ready.

sebastienvas commented:
Are we at a point where we can merge this ?

krzyzacy commented:
rebase and :shipit: ?

sebastienvas commented:
PTAL

krzyzacy commented:
😂 I'm OOO tomorrow, Monday morning?

krzyzacy commented:
get it going :-) 👀 cc @cjwagner 
/lgtm
